src/osd/OSD.cc

```diff
@@ -1436,13 +1436,11 @@ bool OSDService::prepare_to_stop()
 	osdmap->get_epoch(),
 	true  // request ack
 	));
-    utime_t now = ceph_clock_now();
-    utime_t timeout;
-    timeout.set_from_double(now + cct->_conf->osd_mon_shutdown_timeout);
-    while ((ceph_clock_now() < timeout) &&
-       (get_state() != STOPPING)) {
-      is_stopping_cond.WaitUntil(is_stopping_lock, timeout);
-    }
+    ceph::timespan dur = ceph::make_timespan(cct->_conf->osd_mon_shutdown_timeout);
+    cv_status r;
+    do {
+      r = is_stopping_cond.wait_for(l, dur);
+    } while (r == cv_status::no_timeout && get_state() != STOPPING);
```

Comment:
@tchaikov I think it makes more sense to rely on the return value rather than test for time ourselves. The difference now is that `wait_for()` will get called even if `osd_mon_shutdown_timeout` is set to 0 (which wasn't the case before), but will return immediately, achieving the same behavior.

Reply:
@mogeb if we are waiting for something to happen, how about putting something like
```c++
const auto timeout = ceph::make_timespan(cct->_conf->osd_mon_shutdown_timeout
is_stopping_cond.wait_for(l, timeout,
    [this] { return get_state() == STOPPING; });
```

Reply:
@tchaikov updated. i had to add the implement the `wait_for()` method for `condition_variable_debug`

Reply:
thanks!